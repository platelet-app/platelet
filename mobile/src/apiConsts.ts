export const protectedFields = [
    "id",
    "_version",
    "_lastChangedAt",
    "_deleted",
    "updatedAt",
    "createdAt",
    "tenantId",
];

export enum deliverableIcons {
    bug = "BUG",
    child = "CHILD",
    document = "DOCUMENT",
    equipment = "EQUIPMENT",
    other = "OTHER",
}

export enum userRoles {
    admin = "ADMIN",
    user = "USER",
    coordinator = "COORDINATOR",
    rider = "RIDER",
}

export type UserRole = "ADMIN" | "RIDER" | "USER" | "COORDINATOR";

export enum commentVisibility {
    everyone = "EVERYONE",
    me = "ME",
}

export enum tasksStatus {
    new = "NEW",
    active = "ACTIVE",
    pickedUp = "PICKED_UP",
    droppedOff = "DROPPED_OFF",
    rejected = "REJECTED",
    cancelled = "CANCELLED",
    abandoned = "ABANDONED",
    completed = "COMPLETED",
    pending = "PENDING",
}

export type TaskTimeName =
    | "timePickedUpSenderName"
    | "timeDroppedOffRecipientName"
    | null;

export type TaskTimeKey =
    | "timeOfCall"
    | "timePickedUp"
    | "timeDroppedOff"
    | "timeCancelled"
    | "timeRejected"
    | "timeRiderHome"
    | "timePickedUpSenderName"
    | "timeDroppedOffRecipientName"
    | null;

export enum priorities {
    high = "HIGH",
    medium = "MEDIUM",
    low = "LOW",
}

export enum deliverableUnits {
    none = "NONE",
    liter = "LITER",
    milliliter = "MILLILITER",
    gram = "GRAM",
    item = "ITEM",
    box = "BOX",
}

export enum S3ObjectAccessLevels {
    protected = "protected",
    public = "public",
    private = "private",
}
