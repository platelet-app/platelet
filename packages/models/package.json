{
    "name": "@platelet-app/models",
    "version": "1.0.1",
    "description": "DataStore models for the platelet.app project",
    "homepage": "https://platelet.app",
    "bugs": {
        "url": "https://github.com/platelet-app/platelet/issues"
    },
    "repository": {
        "type": "git",
        "url": "git+https://github.com/platelet-app/platelet.git"
    },
    "license": "Apache-2.0",
    "author": "Theo Cranmore <theo@platelet.app>",
    "type": "module",
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "scripts": {
        "build": "cd ../.. && amplify api gql-compile && amplify codegen models && cd packages/models && cp ../../models/* dist",
        "prepublishOnly": "npm version patch"
    }
}
