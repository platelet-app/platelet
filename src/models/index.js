// @ts-check
import { initSchema } from '@aws-amplify/datastore';
import { schema } from './schema';

const Role = {
  "USER": "USER",
  "COORDINATOR": "COORDINATOR",
  "RIDER": "RIDER",
  "ADMIN": "ADMIN"
};

const CommentVisibility = {
  "EVERYONE": "EVERYONE",
  "ME": "ME"
};

const TaskStatus = {
  "NEW": "NEW",
  "ACTIVE": "ACTIVE",
  "PICKED_UP": "PICKED_UP",
  "DROPPED_OFF": "DROPPED_OFF",
  "CANCELLED": "CANCELLED",
  "REJECTED": "REJECTED",
  "ABANDONED": "ABANDONED",
  "COMPLETED": "COMPLETED"
};

const Priority = {
  "HIGH": "HIGH",
  "MEDIUM": "MEDIUM",
  "LOW": "LOW"
};

const DeliverableTypeIcon = {
  "BUG": "BUG",
  "CHILD": "CHILD",
  "DOCUMENT": "DOCUMENT",
  "EQUIPMENT": "EQUIPMENT",
  "OTHER": "OTHER"
};

const DeliverableUnit = {
  "NONE": "NONE",
  "LITER": "LITER",
  "MILLILITER": "MILLILITER",
  "GRAM": "GRAM",
  "ITEM": "ITEM",
  "BOX": "BOX"
};

const { User, PossibleRiderResponsibilities, RiderResponsibility, Comment, TaskAssignee, Task, Location, Handover, Deliverable, DeliverableType, VehicleAssignment, Vehicle, Tenant, Statistics, SendFeedback, AddressAndContactDetails, S3Object } = initSchema(schema);

export {
  User,
  PossibleRiderResponsibilities,
  RiderResponsibility,
  Comment,
  TaskAssignee,
  Task,
  Location,
  Handover,
  Deliverable,
  DeliverableType,
  VehicleAssignment,
  Vehicle,
  Tenant,
  Role,
  CommentVisibility,
  TaskStatus,
  Priority,
  DeliverableTypeIcon,
  DeliverableUnit,
  Statistics,
  SendFeedback,
  AddressAndContactDetails,
  S3Object
};