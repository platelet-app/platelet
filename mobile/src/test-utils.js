import React from "react";
import createSagaMiddleware from "redux-saga";
import { NavigationContainer } from "@react-navigation/native";
import * as models from "./models";
import { Provider, useDispatch } from "react-redux";
import { render as rtlRender } from "@testing-library/react-native";
import { initialiseApp } from "./redux/initialise/initialiseActions";
import rootSaga from "./redux/RootSagas";
import { configureStore } from "@reduxjs/toolkit";
import rootReducer from "./redux/Reducers";
import { PaperProvider } from "react-native-paper";

const testUserModel = new models.User({
    name: "whoami",
    displayName: "Mock User",
    roles: Object.values(models.Role),
    dateOfBirth: null,
    profilePictureThumbnailURL: null,
});

export const testUser = { ...testUserModel, id: "whoami" };

const sagaOptions = {
    onError: (action, error) => {
        console.log("An uncaught exception has occurred in redux-saga:");
        console.log(action);
        if (error) {
            console.log(error.message);
        }
        throw error;
    },
};

function TestApp(props) {
    const dispatch = useDispatch();
    function initialise() {
        return;
        dispatch(initialiseApp());
    }
    React.useEffect(initialise, []);

    return <>{props.children}</>;
}

const sagaMiddleWare = createSagaMiddleware(sagaOptions);

function render(
    ui,
    {
        preloadedState,
        store = configureStore({
            reducer: rootReducer,
            middleware: [sagaMiddleWare],
            preloadedState: {
                roleView: "ALL",
                awsHubDataStoreEventsReducer: {
                    networkStatus: true,
                    ready: true,
                },
                ...preloadedState,
            },
        }),
        ...renderOptions
    } = {}
) {
    sagaMiddleWare.run(rootSaga);
    function Wrapper(props) {
        return (
            <Provider store={store}>
                <PaperProvider>
                    <NavigationContainer>{props.children}</NavigationContainer>
                </PaperProvider>
            </Provider>
        );
    }
    return {
        store,
        component: rtlRender(ui, { wrapper: Wrapper, ...renderOptions }),
    };
}
// re-export everything
export * from "@testing-library/react-native";
// override render method
export { render };
