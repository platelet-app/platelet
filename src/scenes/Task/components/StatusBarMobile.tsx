import React, { useEffect, useRef, useState } from "react";
import Typography from "@mui/material/Typography";
import { AppBar, Chip, Stack } from "@mui/material";
import { ArrowButton } from "../../../components/Buttons";
import { taskStatusHumanReadable } from "../../../utilities";
import useMediaQuery from "@mui/material/useMediaQuery";
import { useTheme } from "@mui/material/styles";
import { makeStyles } from "tss-react/mui";
import IconButton from "@mui/material/IconButton";
import { useDispatch, useSelector } from "react-redux";
import { displayErrorNotification } from "../../../redux/notifications/NotificationsActions";
import { DataStore } from "aws-amplify";
import * as models from "../../../models";
import { dataStoreModelSyncedStatusSelector } from "../../../redux/Selectors";
import generateClipboardTextFromTask from "../../../utilities/generateClipboardTextFromTask";
import { copyStringToClipboard } from "../../../utilities/copyStringToClipboard";
import CopyFailedDialog from "../../../components/CopyFailedDialog";
const colourBarPercent = "90%";

const generateClass = (theme: any, status: models.TaskStatus) => {
    if (status) {
        return {
            padding: 2,
            display: "flex",
            width: "100%",
            paddingLeft: 15,
            paddingRight: 15,
            italic: {
                fontStyle: "italic",
            },
            background: `linear-gradient(0deg,
        ${theme.palette.background.paper}
        ${colourBarPercent},
        ${theme.palette.background.paper}
        ${colourBarPercent},
        ${theme.palette.taskStatus[status]}
        ${colourBarPercent},
        ${theme.palette.taskStatus[status]} 100%)`,
        };
    } else {
        return {
            padding: 2,
            display: "flex",
            width: "100%",
            paddingLeft: 15,
            paddingRight: 15,
            italic: {
                fontStyle: "italic",
            },
            background:
                theme.palette.mode === "dark"
                    ? theme.palette.background.paper
                    : theme.palette.primary.main,
        };
    }
};

// @ts-ignore
const useStyles = makeStyles()((theme, { status }) => {
    const root = generateClass(theme, status);
    return {
        root,
        statusText: {
            fontWeight: "bold",
            color: theme.palette.mode === "dark" ? "white" : "black",
        },
        items: {
            marginTop: 5,
        },
    };
});

type StatusBarProps = {
    handleClose?: () => void;
    taskId?: string;
};

const StatusBarMobile: React.FC<StatusBarProps> = ({ taskId, handleClose }) => {
    const [copied, setCopied] = useState<null | Boolean>(null);
    const [copyText, setCopyText] = useState<string | null>("");
    const theme = useTheme();
    const isSm = useMediaQuery(theme.breakpoints.down("md"));
    const dispatch = useDispatch();
    const taskModelsSynced = useSelector(
        dataStoreModelSyncedStatusSelector
    ).Task;
    const taskObserver = useRef({ unsubscribe: () => {} });
    const [status, setStatus] = useState<models.TaskStatus | null>(null);
    // @ts-ignore
    const { classes } = useStyles({ status });

    const getTask = React.useCallback(async (taskId) => {
        try {
            const task = await DataStore.query(models.Task, taskId);
            taskObserver.current = DataStore.observe(
                models.Task,
                taskId
            ).subscribe(({ element }) => {
                setStatus(element.status as models.TaskStatus);
            });
            if (task) setStatus(task.status as models.TaskStatus);
        } catch (e) {
            console.log(e);
        }
    }, []);
    useEffect(() => {
        getTask(taskId);
    }, [taskId, taskModelsSynced, getTask]);
    async function copyToClipboard() {
        if (!taskId) {
            dispatch(displayErrorNotification("Copy failed."));
            return;
        }
        try {
            const taskResult = await DataStore.query(models.Task, taskId);
            if (!taskResult) throw new Error("Task not found.");
            const deliverablesResult = await DataStore.query(
                models.Deliverable
            );
            const deliverables = deliverablesResult.filter(
                (d) => d.task && d.task.id === taskResult.id
            );
            const result = { ...taskResult, deliverables };
            const textToCopy = generateClipboardTextFromTask(result);
            try {
                copyStringToClipboard(textToCopy).then(
                    () => {
                        setCopied(true);
                    },
                    () => {
                        setCopyText(textToCopy);
                    }
                );
            } catch (e) {
                setCopyText(textToCopy);
            }
        } catch (e) {
            console.log(e);
            setCopied(false);
        }
    }
    let copyLabel = "Copy to clipboard";
    if (copied !== null && copied) {
        copyLabel = "Copy successful!";
    } else if (copied !== null && !copied) {
        copyLabel = "Copy failed!";
    }
    let copyColor: "success" | "secondary" | "primary" = "primary";
    if (copied !== null && copied) {
        copyColor = "success";
    } else if (copied !== null && !copied) {
        copyColor = "secondary";
    }
    return (
        <>
            <AppBar
                position={isSm ? "relative" : "sticky"}
                className={classes.root}
            >
                <Stack
                    direction={"row"}
                    justifyContent={"space-between"}
                    alignItems={"center"}
                    sx={{ paddingTop: 1, width: "100%" }}
                >
                    <IconButton
                        aria-label={"Close"}
                        size={"small"}
                        onClick={handleClose}
                    >
                        <ArrowButton size={3} direction={"back"} />
                    </IconButton>
                    <Typography
                        data-cy="task-status"
                        className={classes.statusText}
                    >
                        {taskStatusHumanReadable(status)}
                    </Typography>
                    <Chip
                        onClick={copyToClipboard}
                        variant={copied === null ? "outlined" : "filled"}
                        color={copyColor}
                        sx={{ marginRight: isSm ? 0 : 2 }}
                        label={copyLabel}
                    />
                </Stack>
            </AppBar>
            <CopyFailedDialog
                text={copyText || ""}
                onClose={() => setCopyText(null)}
            />
        </>
    );
};

export default StatusBarMobile;
