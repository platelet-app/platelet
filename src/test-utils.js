import React from "react";
import createSagaMiddleware from "redux-saga";
import * as models from "./models";
import {
    ThemeProvider,
    StyledEngineProvider,
    createTheme,
} from "@mui/material/styles";
import { CssBaseline } from "@mui/material";
import { SnackbarProvider, withSnackbar } from "notistack";
import { Provider, useDispatch, useSelector } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import mediaQuery from "css-mediaquery";
import ReactNotification from "react-notifications-component";
import "react-notifications-component/dist/theme.css";
import AdapterDateFns from "@mui/lab/AdapterDateFns";
import LocalizationProvider from "@mui/lab/LocalizationProvider";
import { render as rtlRender } from "@testing-library/react";
import { initialiseApp } from "./redux/initialise/initialiseActions";
import rootSaga from "./redux/RootSagas";
import { DismissButton } from "./styles/common";
import SnackNotificationButtons from "./components/SnackNotificationButtons";
import { configureStore } from "@reduxjs/toolkit";
import rootReducer from "./redux/Reducers";
import { userRoles } from "./apiConsts";

const testUserModel = new models.User({
    name: "whoami",
    displayName: "Mock User",
    roles: Object.values(userRoles),
    dateOfBirth: null,
    profilePictureThumbnailURL: null,
});

export const testUser = { ...testUserModel, id: "whoami" };

const taskStatus = {
    NEW: "rgba(252, 231, 121, 1)",
    ACTIVE: "cornflowerblue",
    PICKED_UP: "orange",
    DROPPED_OFF: "lightgreen",
    CANCELLED: "blue",
    REJECTED: "grey",
};
const theme = createTheme({
    palette: {
        mode: "light",
        background: {
            default: "rgb(235, 235, 235)",
        },
        taskStatus,
    },
});

export function createMatchMedia(width) {
    return (query) => ({
        matches: mediaQuery.match(query, {
            width,
        }),
        addListener: () => {},
        removeListener: () => {},
    });
}
function getRandomInt(min, max) {
    min = Math.ceil(min);
    max = Math.floor(max);
    return Math.floor(Math.random() * (max - min) + min); //The maximum is exclusive and the minimum is inclusive
}

export function generateTimes(previous = null, hours = 2) {
    let date;
    if (previous) date = new Date(previous);
    else date = new Date();
    if (!previous) date.setHours(date.getHours() - hours);
    date.setMinutes(date.getMinutes() + getRandomInt(20, 30));
    const timeOfCall = date.toISOString();
    date.setMinutes(date.getMinutes() + getRandomInt(20, 30));
    const timePickedUp = date.toISOString();
    date.setMinutes(date.getMinutes() + getRandomInt(20, 30));
    const timeDroppedOff = date.toISOString();

    return { timeDroppedOff, timePickedUp, timeOfCall };
}

const sagaOptions = {
    onError: (action, error) => {
        console.log("An uncaught exception has occurred in redux-saga:");
        console.log(action);
        if (error) {
            console.log(error.message);
        }
        throw error;
    },
};

function TestApp(props) {
    const dispatch = useDispatch();
    function initialise() {
        return;
        dispatch(initialiseApp());
    }
    React.useEffect(initialise, []);

    const incomingNotification = useSelector((state) => state.notification);

    const snackDismissAction = (key) => (
        <React.Fragment>
            <DismissButton onClick={() => props.closeSnackbar(key)} />
        </React.Fragment>
    );

    function showNotification() {
        if (incomingNotification) {
            const { message, options, restoreCallback, viewLink } =
                incomingNotification;
            options.action = (key) => (
                <SnackNotificationButtons
                    restoreCallback={restoreCallback}
                    viewLink={viewLink}
                    snackKey={key}
                    closeSnackbar={props.closeSnackbar}
                />
            );
            props.enqueueSnackbar(message, options);
        }
    }

    React.useEffect(showNotification, [incomingNotification]);
    return (
        <StyledEngineProvider injectFirst>
            <ThemeProvider theme={theme}>
                <CssBaseline />
                {props.children}
            </ThemeProvider>
        </StyledEngineProvider>
    );
}

const AppSnacked = withSnackbar(TestApp);

const sagaMiddleWare = createSagaMiddleware(sagaOptions);

function render(
    ui,
    {
        preloadedState,
        store = configureStore({
            reducer: rootReducer,
            middleware: [sagaMiddleWare],
            preloadedState: {
                roleView: "ALL",
                awsHubDataStoreEventsReducer: {
                    networkStatus: true,
                    ready: true,
                },
                ...preloadedState,
            },
        }),
        ...renderOptions
    } = {}
) {
    sagaMiddleWare.run(rootSaga);
    function Wrapper(props) {
        return (
            <Provider store={store}>
                <BrowserRouter>
                    <LocalizationProvider dateAdapter={AdapterDateFns}>
                        <ReactNotification />
                        <SnackbarProvider maxSnack={1}>
                            <AppSnacked {...props}>{props.children}</AppSnacked>
                        </SnackbarProvider>
                    </LocalizationProvider>
                </BrowserRouter>
            </Provider>
        );
    }
    return {
        store,
        component: rtlRender(ui, { wrapper: Wrapper, ...renderOptions }),
    };
}
// re-export everything
export * from "@testing-library/react";
// override render method
export { render };
