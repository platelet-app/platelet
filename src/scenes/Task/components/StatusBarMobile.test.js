import { render } from "../../../test-utils";
import { screen, waitFor } from "@testing-library/react";
import { tasksStatus } from "../../../apiConsts";
import * as models from "../../../models/index";
import userEvent from "@testing-library/user-event";
import { createMatchMedia } from "../../../test-utils";
import { DataStore } from "aws-amplify";
import StatusBarMobile from "./StatusBarMobile";

describe.only("StatusBarMobile", () => {
    beforeAll(() => {
        window.matchMedia = createMatchMedia(window.innerWidth);
    });
    afterEach(() => {
        jest.restoreAllMocks();
    });
    beforeEach(() => {
        window.matchMedia = createMatchMedia(240);
    });
    it("renders correctly", () => {
        render(<StatusBarMobile />);
    });

    it.each`
        status
        ${tasksStatus.completed}
        ${tasksStatus.cancelled}
        ${tasksStatus.pickedUp}
        ${tasksStatus.droppedOff}
        ${tasksStatus.abandoned}
        ${tasksStatus.active}
        ${tasksStatus.new}
        ${tasksStatus.rejected}
    `("renders the correct status", async ({ status }) => {
        const mockTask = await DataStore.save(new models.Task({ status }));
        const querySpy = jest.spyOn(DataStore, "query");
        render(<StatusBarMobile taskId={mockTask.id} />);
        await waitFor(() => {
            expect(querySpy).toHaveBeenCalledWith(models.Task, mockTask.id);
        });
        if (status === tasksStatus.droppedOff) status = "DELIVERED";
        else if (status === tasksStatus.pickedUp) status = "PICKED UP";
        expect(screen.getByText(status)).toBeInTheDocument();
    });

    test("if the copy fails, show it in a dialog instead", async () => {
        Object.assign(window, {
            cordova: undefined,
        });
        Object.assign(window.navigator, {
            clipboard: {
                writeText: jest.fn().mockImplementation(() => Promise.reject()),
            },
        });
        const timeOfCall = "2022-09-14T07:55:07.473Z";
        const pickUpLocation = new models.Location({
            line1: "line one",
            ward: "test ward",
            postcode: "postcode",
        });
        const dropOffLocation = new models.Location({
            line1: "something",
            ward: "some ward",
            postcode: "some postcode",
        });
        const mockTask = new models.Task({
            timeOfCall,
            status: tasksStatus.new,
            pickUpLocation,
            dropOffLocation,
            priority: models.Priority.HIGH,
        });
        await DataStore.save(mockTask);
        const mockDeliverableType = new models.DeliverableType({
            label: "test deliverable",
        });
        await DataStore.save(mockDeliverableType);
        const mockDeliverables = [
            {
                unit: "ITEM",
                count: 1,
                orderInGrid: 0,
                deliverableType: mockDeliverableType,
            },
            {
                unit: "ITEM",
                count: 2,
                orderInGrid: 0,
                deliverableType: mockDeliverableType,
            },
            {
                unit: models.DeliverableUnit.LITER,
                count: 3,
                orderInGrid: 0,
                deliverableType: mockDeliverableType,
            },
        ];
        await Promise.all(
            mockDeliverables.map((deliverable) =>
                DataStore.save(
                    new models.Deliverable({ task: mockTask, ...deliverable })
                )
            )
        );
        const querySpy = jest.spyOn(DataStore, "query");
        render(<StatusBarMobile taskId={mockTask.id} />);
        await waitFor(() => {
            expect(querySpy).toHaveBeenCalledWith(models.Task, mockTask.id);
        });
        const copyButton = screen.getByText("Copy to clipboard");
        userEvent.click(copyButton);
        await screen.findByTestId("copy-failed-dialog");
        expect(
            screen.getByText(
                "TOC: 07:55 FROM: test ward - line one, postcode TO: some ward - something, some postcode PRIORITY: high ITEMS: test deliverable x 1, test deliverable x 2, test deliverable x 3"
            )
        ).toBeInTheDocument();
        userEvent.click(screen.getByTestId("copy-ok-button"));
        await waitFor(() => {
            expect(screen.queryByTestId("copy-failed-dialog")).toBeNull();
        });
    });

    test("click the copy to clipboard button on mobile", async () => {
        Object.assign(window, {
            cordova: {
                plugins: {
                    clipboard: {
                        copy: jest
                            .fn()
                            .mockImplementation(() => Promise.resolve()),
                    },
                },
            },
        });
        const timeOfCall = "2022-09-14T07:55:07.473Z";
        const pickUpLocation = new models.Location({
            line1: "line one",
            ward: "test ward",
            postcode: "postcode",
        });
        const dropOffLocation = new models.Location({
            line1: "something",
            ward: "some ward",
            postcode: "some postcode",
        });
        const mockTask = new models.Task({
            timeOfCall,
            status: tasksStatus.new,
            pickUpLocation,
            dropOffLocation,
            priority: models.Priority.HIGH,
        });
        await DataStore.save(mockTask);
        const mockDeliverableType = new models.DeliverableType({
            label: "test deliverable",
        });
        await DataStore.save(mockDeliverableType);
        const mockDeliverables = [
            {
                unit: "ITEM",
                count: 1,
                orderInGrid: 0,
                deliverableType: mockDeliverableType,
            },
            {
                unit: "ITEM",
                count: 2,
                orderInGrid: 0,
                deliverableType: mockDeliverableType,
            },
            {
                unit: models.DeliverableUnit.LITER,
                count: 3,
                orderInGrid: 0,
                deliverableType: mockDeliverableType,
            },
        ];
        await Promise.all(
            mockDeliverables.map((deliverable) =>
                DataStore.save(
                    new models.Deliverable({ task: mockTask, ...deliverable })
                )
            )
        );
        const querySpy = jest.spyOn(DataStore, "query");
        let cordovaSpy;
        cordovaSpy = jest.spyOn(window.cordova.plugins.clipboard, "copy");
        render(<StatusBarMobile taskId={mockTask.id} />);
        await waitFor(() => {
            expect(querySpy).toHaveBeenCalledWith(models.Task, mockTask.id);
        });
        const copyButton = screen.getByText("Copy to clipboard");
        expect(copyButton).toBeInTheDocument();
        userEvent.click(copyButton);
        await waitFor(() => {
            expect(cordovaSpy).toHaveBeenCalled();
        });
        expect(cordovaSpy).toMatchSnapshot();
        // I don't know why this fails, I can see it is working in the app
        //expect(await screen.findByText("Copy successful!")).toBeInTheDocument();
    });

    test("click the copy to clipboard button", async () => {
        Object.assign(window, {
            cordova: undefined,
        });
        Object.assign(window.navigator, {
            clipboard: {
                writeText: jest
                    .fn()
                    .mockImplementation(() => Promise.resolve()),
            },
        });
        const timeOfCall = "2022-09-14T07:55:07.473Z";
        const pickUpLocation = new models.Location({
            line1: "line one",
            ward: "test ward",
            postcode: "postcode",
        });
        const dropOffLocation = new models.Location({
            line1: "something",
            ward: "some ward",
            postcode: "some postcode",
        });
        const mockTask = new models.Task({
            timeOfCall,
            status: tasksStatus.new,
            pickUpLocation,
            dropOffLocation,
            priority: models.Priority.HIGH,
        });
        await DataStore.save(mockTask);
        const mockDeliverableType = new models.DeliverableType({
            label: "test deliverable",
        });
        await DataStore.save(mockDeliverableType);
        const mockDeliverables = [
            {
                unit: "ITEM",
                count: 1,
                orderInGrid: 0,
                deliverableType: mockDeliverableType,
            },
            {
                unit: "ITEM",
                count: 2,
                orderInGrid: 0,
                deliverableType: mockDeliverableType,
            },
            {
                unit: models.DeliverableUnit.LITER,
                count: 3,
                orderInGrid: 0,
                deliverableType: mockDeliverableType,
            },
        ];
        await Promise.all(
            mockDeliverables.map((deliverable) =>
                DataStore.save(
                    new models.Deliverable({ task: mockTask, ...deliverable })
                )
            )
        );
        const querySpy = jest.spyOn(DataStore, "query");
        const clipboardSpy = jest.spyOn(navigator.clipboard, "writeText");
        render(<StatusBarMobile taskId={mockTask.id} />);
        await waitFor(() => {
            expect(querySpy).toHaveBeenCalledWith(models.Task, mockTask.id);
        });
        const copyButton = screen.getByText("Copy to clipboard");
        expect(copyButton).toBeInTheDocument();
        userEvent.click(copyButton);
        await waitFor(() => {
            expect(clipboardSpy).toHaveBeenCalled();
        });
        expect(clipboardSpy).toMatchSnapshot();
        expect(await screen.findByText("Copy successful!")).toBeInTheDocument();
    });

    it("fails to copy task data to clipboard", async () => {
        jest.restoreAllMocks();
        render(<StatusBarMobile taskId={"nope"} />);
        const copyButton = screen.getByText("Copy to clipboard");
        expect(copyButton).toBeInTheDocument();
        userEvent.click(copyButton);
        expect(await screen.findByText("Copy failed!")).toBeInTheDocument();
    });

    test("click the back button on mobile", async () => {
        const mockClose = jest.fn();
        render(<StatusBarMobile handleClose={mockClose} />);
        const closeButton = screen.getByRole("button", { name: "Close" });
        expect(closeButton).toBeInTheDocument();
        userEvent.click(closeButton);
        expect(mockClose).toHaveBeenCalled();
    });
});
