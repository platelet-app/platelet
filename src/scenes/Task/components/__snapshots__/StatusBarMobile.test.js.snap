// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusBarMobile click the copy to clipboard button 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "TOC: 07:55 FROM: test ward - line one, postcode TO: some ward - something, some postcode PRIORITY: high ITEMS: test deliverable x 1, test deliverable x 2, test deliverable x 3",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`StatusBarMobile click the copy to clipboard button on mobile 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "TOC: 07:55 FROM: test ward - line one, postcode TO: some ward - something, some postcode PRIORITY: high ITEMS: test deliverable x 1, test deliverable x 2, test deliverable x 3",
      [Function],
      [Function],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
